### PR TITLE
Restructure how we decide what functions to generate methods for

### PR DESCRIPTION
**Pull request by jesyspa**
As it is, we treat certain inline functions as non-inline but don't generate the corresponding methods.

I tried creating a test for this but it requires some access to the standard library that is hard to set up.

### CODE REVIEW

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/callables/InlineNamedFunction.kt [hunk 1]
```diff
@@ -26,4 +26,6 @@ class InlineNamedFunction(
         val paramNames = symbol.valueParameterSymbols.map { it.name }
         return ctx.insertInlineFunctionCall(signature, paramNames, args, body)
     }
+
+    override fun toViperMethod(): Nothing? = null
```
Comment (gabryon99) [inline plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/callables/InlineNamedFunction.kt:30]: Why `Nothing?` as return type and not `Method?`?

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/ProgramConverter.kt [hunk 2]
```diff
@@ -75,8 +75,9 @@ class ProgramConverter(val session: FirSession, override val config: PluginConfi
     override fun embedFunction(symbol: FirFunctionSymbol<*>): FunctionEmbedding =
         methods.getOrPut(symbol.embedName(this)) {
             val signature = embedFullSignature(symbol)
-            val embedding = UserFunctionEmbedding(processCallable(symbol, signature))
-            embedding.viperMethod = convertMethodWithoutBody(symbol, signature)
+            val callable = processCallable(symbol, signature)
+            val embedding = UserFunctionEmbedding(callable)
+            embedding.viperMethod = callable.toViperMethod()
```
Comment (gabryon99) [inline plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/ProgramConverter.kt:80]: Wouldn't be possible initializing the field `viperMethod` within an `init` block in `UserFunctionEmbedding` class?

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/callables/InlineNamedFunction.kt (hunk 1 above)
Comment (jesyspa) [reply plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/callables/InlineNamedFunction.kt:30]: `Nothing?` is more specific; we guarantee that inline methods do not generate a method.

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/ProgramConverter.kt (hunk 2 above)
Comment (jesyspa) [reply plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/ProgramConverter.kt:80]: This is not the only place `viperMethod` is set; `UserFunctionEmbedding` does not know whether the method should be set based on the `callable` passed in or not.

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/callables/InlineNamedFunction.kt (hunk 1 above)
Comment (jesyspa) [reply plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/callables/InlineNamedFunction.kt:30]: We can, `Nothing?` is a subtype of `Method?`, hence every value of type `Nothing?` (i.e. `null`) also has type `Method?`.  But `Nothing?` is more specific; e.g. someone doing `val m = toViperMethod()` and then trying `m as? Method` will receive a warning that this cast will never succeed, while if we return `Method?` we do not have a corresponding guarantee.

Comment (gabryon99) [reply plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/callables/InlineNamedFunction.kt:30]: Ok, thank you for the clarification!

**Review by gabryon99 — APPROVED** [review]

**Pull request merged**